sql/sql_admin.cc

```diff
@@ -1608,18 +1608,45 @@ bool Sql_cmd_optimize_table::execute(THD *thd)
                          FALSE, UINT_MAX, FALSE))
     goto error; /* purecov: inspected */
 
-  WSREP_TO_ISOLATION_BEGIN_WRTCHK(NULL, NULL, first_table);
+#ifdef WITH_WSREP
+  if (WSREP(thd) &&
+      (!thd->is_current_stmt_binlog_format_row() ||
+       !thd->find_temporary_table(first_table)))
+  {
+    /*
+      It makes sense to set auto_increment_* to defaults in TOI operations.
+      Must be done before wsrep_TOI_begin() since Query_log_event encapsulating
+      TOI statement and auto inc variables for wsrep replication is constructed
+      there. Variables are reset back in THD::reset_for_next_command() before
+      processing of next command.
+    */
+    if (wsrep_auto_increment_control)
+    {
+      thd->variables.auto_increment_offset = 1;
+      thd->variables.auto_increment_increment = 1;
+    }
+
+    wsrep::key_array keys;
+    if (wsrep_thd_is_local(thd) &&
+        !wsrep_append_fk_parent_table(thd, first_table, &keys))
```

Comment:
If wsrep_append_fk_parent_table fails, then we skip TOI, and but still perform the OPTIMIZE locally.
This is likely harmless. Still, I would prefer to perform OPTIMIZE only if we did enter TOI mode.

Reply:
Yes, it is certainly possible to change the code so that `wsrep_append_fk_parent_table()` failure causes `Sql_cmd_optimize_table::execute()` to fail.
But the current code is now analogous to the implementation of `Sql_cmd_alter_table::execute()`.